src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers/ClientMethodsAnalyzer.cs

```diff
@@ -21,104 +28,126 @@ public class ClientMethodsAnalyzer : ClientAnalyzerBase
             Descriptors.AZC0015
         });
 
-        private static void CheckClientMethod(ISymbolAnalysisContext context, IMethodSymbol member)
+        private static void CheckCancellationTokenOrRequestContext(ISymbolAnalysisContext context, IMethodSymbol member)
         {
-            static bool SupportsCancellationsParameter(IParameterSymbol parameterSymbol)
+            static bool IsRequestContext(IParameterSymbol parameterSymbol)
             {
-                return (parameterSymbol.Name == "cancellationToken" && parameterSymbol.Type.Name == "CancellationToken") ||
-                       (parameterSymbol.Name == "context" && parameterSymbol.Type.Name == "RequestContext");
+                return parameterSymbol.Name == "context" && parameterSymbol.Type.Name == "RequestContext";
             }
 
-            CheckClientMethodReturnType(context, member);
+            static bool IsCancellationToken(IParameterSymbol parameterSymbol)
+            {
+                return parameterSymbol.Name == "cancellationToken" && parameterSymbol.Type.Name == "CancellationToken";
+            }
 
-            if (!member.IsVirtual && !member.IsOverride)
+            static bool IsCancellationOrRequestContext(IParameterSymbol parameterSymbol)
             {
-                context.ReportDiagnostic(Diagnostic.Create(Descriptors.AZC0003, member.Locations.First()), member);
+                return IsCancellationToken(parameterSymbol) || IsRequestContext(parameterSymbol);
             }
 
             var lastArgument = member.Parameters.LastOrDefault();
-            var supportsCancellations = lastArgument != null && SupportsCancellationsParameter(lastArgument);
+            var isCancellationOrRequestContext = lastArgument != null && IsCancellationOrRequestContext(lastArgument);
 
-            if (!supportsCancellations)
+            if (!isCancellationOrRequestContext)
             {
                 var overloadSupportsCancellations = FindMethod(
                     member.ContainingType.GetMembers(member.Name).OfType<IMethodSymbol>(),
                     member.TypeParameters,
                     member.Parameters,
-                    p => SupportsCancellationsParameter(p));
+                    p => IsCancellationToken(p));
 
-                if (overloadSupportsCancellations != null)
+                if (overloadSupportsCancellations == null)
                 {
-                    // Skip methods that have overloads with cancellation tokens
-                    return;
+                    context.ReportDiagnostic(Diagnostic.Create(Descriptors.AZC0002, member.Locations.FirstOrDefault()), member);
                 }
-
-                context.ReportDiagnostic(Diagnostic.Create(Descriptors.AZC0002, member.Locations.FirstOrDefault()), member);
             }
-            else if (!lastArgument.IsOptional)
+            else if (IsCancellationToken(lastArgument))
```

Comment:
Can we collapse the nested `if` statements here, since it appears there is not alternate branch for either?
```suggestion
            else if (IsCancellationToken(lastArgument) && !lastArgument.IsOptional)
```